materialize-mongodb/driver.go

```diff
@@ -0,0 +1,229 @@
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = d.Connect(ctx, cfg)
+	if err != nil {
+		return nil, fmt.Errorf("connecting to database: %w", err)
+	}
+
+	return &pm.ApplyResponse{
+		ActionDescription: "",
+	}, nil
+}
+
+// ApplyDelete is no-op as well, since this operation is going to be superseded
+// by an "Apply" operation with no bindings specified
+func (d driver) ApplyDelete(ctx context.Context, req *pm.ApplyRequest) (*pm.ApplyResponse, error) {
+	if err := req.Validate(); err != nil {
+		return nil, fmt.Errorf("validating request: %w", err)
+	}
+
+	cfg, err := resolveEndpointConfig(req.Materialization.EndpointSpecJson)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = d.Connect(ctx, cfg)
+	if err != nil {
+		return nil, fmt.Errorf("connecting to database: %w", err)
+	}
+
+	return &pm.ApplyResponse{
+		ActionDescription: "",
+	}, nil
+}
+
+
+func (d driver) Transactions(stream pm.Driver_TransactionsServer) error {
+	return pm.RunTransactions(stream, func(ctx context.Context, open pm.TransactionRequest_Open) (pm.Transactor, *pm.TransactionResponse_Opened, error) {
+		var cfg, err = resolveEndpointConfig(open.Materialization.EndpointSpecJson)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		client, err := d.Connect(ctx, cfg)
+		if err != nil {
+			return nil, nil, fmt.Errorf("connecting to database: %w", err)
+		}
+
+		var bindings []*binding
+		for _, b := range open.Materialization.Bindings {
+			res, err := resolveResourceConfig(b.ResourceSpecJson)
+			if err != nil {
+				return nil, nil, err
+			}
+			var collection = client.Database(res.Database).Collection(res.Collection)
+
+			bindings = append(bindings, &binding{
+				collection: collection,
+				res: res,
+			})
+		}
+
+		var fenceCollection = client.Database(cfg.Database).Collection(fenceCollectionName)
+
+		var materialization = string(open.Materialization.Materialization)
+		var filter = bson.D{{
+			"materialization", bson.D{{"$eq", materialization}},
+		}}
+
+		var fence fenceRecord
+		if err := fenceCollection.FindOne(ctx, filter, options.FindOne()).Decode(&fence); err != nil && err != mongo.ErrNoDocuments {
+			return nil, nil, fmt.Errorf("finding existing fence: %w", err)
+		} else if err == mongo.ErrNoDocuments {
+			fence.Materialization = materialization
+			if _, err = fenceCollection.InsertOne(ctx, fence, options.InsertOne()); err != nil {
+				return nil, nil, fmt.Errorf("inserting new fence: %w", err)
+			}
+		}
+
+		var bump = bson.D{{"$inc", bson.D{{"fence", 1}}}}
+		var updateOpts = options.FindOneAndUpdate().SetReturnDocument(options.After)
+		if err = fenceCollection.FindOneAndUpdate(ctx, filter, bump, updateOpts).Decode(&fence); err != nil {
+			return nil, nil, fmt.Errorf("bumping fence: %w", err)
+		}
+
+		return &transactor{
+			materialization: materialization,
+			client: client,
+			bindings: bindings,
+			fenceCollection: fenceCollection,
+			fence: &fence,
+		}, &pm.TransactionResponse_Opened{}, nil
```

Comment:
For an authoritative remote store materialization (which it looks like MongoDB transactions will support, which is cool) there needs to be a runtime checkpoint returned here. Otherwise the runtime will continue from its last checkpoint in the recovery log. In that case there is a chance that the commit future that was returned in the transactor `Store` phase failed to actually commit. This would result in missing data from the endpoint.

I think this could be accomplished by storing the runtime checkpoint with the `fenceRecord` as well. Then when the fence record is loaded here, the last persisted runtime checkpoint in the endpoint can be returned with the opened response.